presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatStreamReader.java

```diff
@@ -0,0 +1,429 @@
+
+    private boolean rowGroupOpen;
+
+    public MapFlatStreamReader(StreamDescriptor streamDescriptor, DateTimeZone hiveStorageTimeZone, AggregatedMemoryContext systemMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "stream is null");
+        this.hiveStorageTimeZone = requireNonNull(hiveStorageTimeZone, "hiveStorageTimeZone is null");
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        this.keyOrcType = streamDescriptor.getNestedStreams().get(0).getStreamType();
+        this.baseValueStreamDescriptor = streamDescriptor.getNestedStreams().get(1);
+    }
+
+    @Override
+    public void prepareNextRead(int batchSize)
+    {
+        readOffset += nextBatchSize;
+        nextBatchSize = batchSize;
+    }
+
+    @Override
+    public Block readBlock(Type type)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        if (readOffset > 0) {
+            if (presentStream != null) {
+                // skip ahead the present bit reader, but count the set bits
+                // and use this as the skip size for the data reader
+                readOffset = presentStream.countBitsSet(readOffset);
+            }
+            if (readOffset > 0) {
+                for (int i = 0; i < valueStreamReaders.size(); i++) {
+                    int valueReadOffset = inMapStreams.get(i).countBitsSet(readOffset);
+                    valueStreamReaders.get(i).prepareNextRead(valueReadOffset);
+                }
+            }
+        }
+
+        boolean[] nullVector = null;
+
+        if (presentStream == null) {
+            for (int i = 0; i < inMapStreams.size(); i++) {
+                if (inMapStreams.get(i) != null) {
+                    inMapVectors[i] = new boolean[nextBatchSize];
+                    inMapStreams.get(i).getSetBits(nextBatchSize, inMapVectors[i]);
+                }
+            }
+        }
+        else {
+            nullVector = new boolean[nextBatchSize];
+            int nullValues = presentStream.getUnsetBits(nextBatchSize, nullVector);
+            if (nullValues != nextBatchSize) {
+                for (int i = 0; i < inMapStreams.size(); i++) {
+                    if (inMapStreams.get(i) != null) {
+                        inMapVectors[i] = new boolean[nextBatchSize];
+                        inMapStreams.get(i).getSetBits(nextBatchSize, inMapVectors[i], nullVector);
+                    }
+                }
+            }
+        }
+
+        MapType mapType = (MapType) type;
+        Type valueType = mapType.getValueType();
+
+        Block[] valueBlocks = new Block[valueStreamReaders.size()];
+
+        for (int i = 0; i < valueStreamReaders.size(); i++) {
+            int mapsContainingKey = 0;
+
+            for (int j = 0; j < nextBatchSize; j++) {
+                mapsContainingKey += inMapVectors[i] == null || inMapVectors[i][j] ? 1 : 0;
```

Comment:
So, if `inMapVectors[i] == null`, everything is considered as true?

Reply:
It's similar to present streams